Smarter differentiable subgraph slicing

zou3519 opened the pull request:
### Motivation

If any inputs require_grad then the graph executor does differential subgraph slicing. The existing algorithm combines adjacent differentiable Node*.

There are two major motivations. The first is improving fusion opportunities: the graph fusion pass runs after differential subgraph slicing. This means that only nodes that are a part of the same differential subgraph may be considered for fusion. If something like the following happens,
```
y = f(x)
k = not_differentiable_op(m)
z = g(y)
```
and f and g are both fusible and differentiable operations, then they will be inserted into different differential subgraphs and not fused together.

The second is to enable JIT optimizations on backward passes for things like an (automatically) unrolled LSTM. Right now, in an unrolled LSTM, we see something like the following:
```
lstm_cell()
non_differentiable_list_op()
lstm_cell()
non_differentiable_list_op()
lstm_cell()
non_differentiable_list_op()
```
Each lstm_cell itself is differentiable and gets put into a separate differential subgraph. During the backwards pass, each prim::DifferentiableSubgraph has its own graph executor: these graph executors cannot talk to each other. It is better if we combined all of the lstm_cells (where applicable) into one differential subgraph so their backward passes are combined into one graph executor that can perform better optimizations than several separate graph executors.

### Problem Statement
Think about the computation graph as a DAG where edges are data dependencies and vertices are operations (the nodes). Each vertex is either black or red; a vertex is colored black if it is differentiable and red otherwise. The goal is to contract edges (merge nodes) to have the fewest black vertices remaining such that the graph is still a DAG.

### Implementation
The algorithm is the following:
- Take the Graph& and create a shadow "DynamicDAG" object to wrap Node* and edges. Each Vertex holds multiple Node* (but starts out holding one Node*) and each edge is a data dependency.
- Greedily contract vertices in the DynamicDAG if they are "differentiable". This operation is unrelated to the Graph&.
  - A Vertex is "differentiable" if all the nodes it holds is differentiable.
  - When contracting vertices, combine their Node* contents.
  - The DynamicDAG keeps its vertices in topological order and complains if the contraction is invalid so everything is good.
- Take the DynamicDAG: reorder the nodes in the Graph& to match the topological order in the DynamicDAG.
- Finally, go through each Vertex in the DynamicDAG: if it contains multiple Node* then merge all of them into a prim::DifferentiableGraph.

The DynamicDAG is based off of the dynamic top sort algorithm in [this paper](https://www.doc.ic.ac.uk/~phjk/Publications/DynamicTopoSortAlg-JEA-07.pdf) by Pearce and Kelly.

### Analysis
Each contractEdge(producer, consumer) call is `O(|AR| log |AR| * min(|out_edges(producer)|, |in_edges(consumer)|)` where `AR` is the "affected region" (defined as the set of nodes that, in topological order, are between producer and consumer). By only considering contractions such that `|ord(producer) - ord(consumer)| < threshold1` and `|out_edges(producer)| < threshold2` we can make each contractEdge(producer, consumer) call take constant time. The resulting algorithm is linear in the number of nodes.

### Test Plan
Added a lot of small test cases.

Looking for suggestions on the following:
- what big computation graphs should I run this on to test how fast or slow it is?
- what things other than correctness should I be thinking about when I test this?

cc @apaszke @zdevito

zou3519 committed:
Smarter differentiable subgraph slicing

If any inputs require_grad then the graph executor does differential subgraph slicing. The existing algorithm combines adjacent differentiable Node*.

There are two major motivations. The first is improving fusion opportunities: the graph fusion pass runs after differential subgraph slicing. This means that only nodes that are a part of the same differential subgraph may be considered for fusion. If something like the following happens,
```
y = f(x)
k = not_differentiable_op(m)
z = g(y)
```
and f and g are both fusible and differentiable operations, then they will be inserted into different differential subgraphs and not fused together.

The second is to enable JIT optimizations on backward passes for things like an (automatically) unrolled LSTM. Right now, in an unrolled LSTM, we see something like the following:
```
lstm_cell()
non_differentiable_list_op()
lstm_cell()
non_differentiable_list_op()
lstm_cell()
non_differentiable_list_op()
```
Each lstm_cell itself is differentiable and gets put into a separate differential subgraph. During the backwards pass, each prim::DifferentiableSubgraph has its own graph executor: these graph executors cannot talk to each other. It is better if we combined all of the lstm_cells (where applicable) into one differential subgraph so their backward passes are combined into one graph executor that can perform better optimizations than several separate graph executors.

Think about the computation graph as a DAG where edges are data dependencies and vertices are operations (the nodes). Each vertex is either black or red; a vertex is colored black if it is differentiable and red otherwise. The goal is to contract edges (merge nodes) to have the fewest black vertices remaining such that the graph is still a DAG.

The algorithm is the following:
- Take the Graph& and create a shadow "DynamicDAG" object to wrap Node* and edges. Each Vertex holds multiple Node* (but starts out holding one Node*) and each edge is a data dependency.
- Greedily contract vertices in the DynamicDAG if they are "differentiable". This operation is unrelated to the Graph&.
  - A Vertex is "differentiable" if all the nodes it holds is differentiable.
  - When contracting vertices, combine their Node* contents.
  - The DynamicDAG keeps its vertices in topological order and complains if the contraction is invalid so everything is good.
- Take the DynamicDAG: reorder the nodes in the Graph& to match the topological order in the DynamicDAG.
- Finally, go through each Vertex in the DynamicDAG: if it contains multiple Node* then merge all of them into a prim::DifferentiableGraph.

The DynamicDAG is based off of the dynamic top sort algorithm in [this paper](https://www.doc.ic.ac.uk/~phjk/Publications/DynamicTopoSortAlg-JEA-07.pdf) by Pearce and Kelly.

Each contractEdge(producer, consumer) call is `O(|AR| * |out_edges(producer)|)` where `AR` is the "affected region" (defined as the set of nodes that, in topological order, are between producer and consumer). By only considering contractions such that `|ord(producer) - ord(consumer)| < threshold1` and `|out_edges(producer)| < threshold2` we can make each contractEdge(producer, consumer) call take constant time. The resulting algorithm is linear in the number of nodes.

WIP. I'm looking for suggestions for the following:
- what big computation graphs should I run this on to test how fast or slow it is?
- what things other than correctness should I be thinking about when I test this?

apaszke commented on test/test_jit.py:
```diff
@@ -62,6 +62,10 @@
 
 
+def pyfn(a, b):
+    return a * b
```
Can this be made local to a test? That might really end up being clearer (or you can make it a static method of the test class).

zou3519 replied:
Something strange happened the last time I made it local to a test, but I'll try again and file an issue if it errors out.

zou3519 replied:
https://github.com/pytorch/pytorch/issues/12393 The scoping behavior on torch.jit.script is a little weird so I couldn't make it local to a test.

Making it a static method of the test class doesn't work either (the compiler complains about "not being able to access members of TestAutodiffSubgraphSlicing")

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+  size_t ord; // index in topological ordering. Is unique.
+  bool visited; // If this vertex has been visited
+  vertex_set<T> in_edges;
+  vertex_set<T> out_edges;
```
We don't really have very large fan in and outs in our nodes, so a vector might be faster here (and would use fewer allocations - every element you insert into an unordered map is a new allocation)

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+
+template <typename T>
+struct Vertex {
+  std::vector<T> rdata; // stored in reverse order
```
In reverse order wrt... what order?

zou3519 replied:
Reverse order to the order in which the Node* appear in the Graph&.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+  ~DynamicDAG();
+
+  Vertex<T>* newVertex(T datum);
+  std::pair<vertex_set<T>,vertex_set<T>> removeVertex(Vertex<T>* v);
```
Can we please make a helper struct for this (think of it as a named tuple). Otherwise it's easy to mix up those arguments.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+
+  // Store vertices indexed by their topological order.
+  // If a vertex v has ord 5, then it can be found at vertices_[5].
+  // There may be gaps in vertices_; this is to enable fast deletion.
```
We don't care about insertion, because this assumes that we'll merge nodes into groups that execute atomically, right?

apaszke replied:
Also, can't we simply store `at::optional<Vertex<T>>`? There's no reason for the heap indirection - this vector can provide you with all the lifetime management you need (it also improves data locality, because they're not scattered all over the heap).

zou3519 replied:
The algorithm reorders`Vertex<T>*` in the `std::vector<at::optional<Vertex<T>*>>`.

If it were `std::vector<at::optional<Vertex<T>>>` instead, reordering the Vertex<T> would involve copies of things larger than pointers, which is probably less performant. Let me know if that is not a correct statement.

apaszke replied:
Yeah that makes sense. It would be great if we could make them `unique_ptr`s!

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+  auto* vertex = new Vertex<T>();
+  vertex->rdata.push_back(datum);
+  vertex->visited = false;
+  vertex->ord = vertices_.size();
```
Can we make a constructor for `Vertex` instead? That would make it harder to forget to initialize a member.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+    }
+    for (auto* v : vertex->out_edges) {
+      JIT_ASSERT(v->ord > ord);
+    }
```
Wouldn't it be sufficient to only check `in_edges` or `out_edges`, since every edge is supposed to be in corresponding sets of its endpoints? It might be a good idea that this is the case, and then only do a single check.

zou3519 replied:
I had some memory corruption earlier from modifying an unordered_set while iterating through it that made it so that some edges were not recorded correctly so this function was helpful to debug that.

`checkInvariants` isn't called anywhere right now, I think I should rename it to `debugCheckInvariants` and use it in test_jit.cpp or something.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+  auto in_edges = std::move(v->in_edges);
+  auto out_edges = std::move(v->out_edges);
+  vertices_[v->ord] = at::nullopt;
+  delete v;
```
No manual memory management like that pls.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+//
+// contractEdge is O(|AR| * |out_edges(producer)|)
+template <typename T>
+bool DynamicDAG<T>::contractEdge(Vertex<T>* producer, Vertex<T>* consumer) {
```
Does edge contraction mean that we fuse/combine producer and consumer into a single node?

zou3519 replied:
Yes

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+  removeEdge(producer, consumer);
+
+  // We can either merge producer into consumer or the other way around.
+  // I've chosen to merge producer into consumer for now but this is arbitrary.
```
If I understand correctly, this choice determines if it's the `in_edges` or `out_edges` insertion that's free, right? We might simple choose the strategy based on their sizes. We sometimes have nodes with large fan in, or large fan out, but rarely with both.

zou3519 replied:
Yes, that is correct. The strategy selection is a great idea.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+
+    for (auto* next : vertex->out_edges) {
+      if (vertex != producer && next->ord == upper_bound) {
+        clearVisited(visited);
```
Can we have a RAII-style struct that calls `clearVisited`? We want to do that on every single exit path from this function, so it seems like a reasonable thing to do.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+        return false;
+      }
+      if (next->ord < upper_bound) {
+        stack.push_back(next);
```
It's faster if you move the `visited` check to here and avoid pushing those node altogether (otherwise, set it to true and push).

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+}
+
+template <typename T>
+bool DynamicDAG<T>::contractionProducesCycle(Vertex<T>* producer, Vertex<T>* consumer) {
```
I feel like this is slightly unnecessary. Couldn't we simply do the `forwardDFS`/`backwardDFS`, and throw if we reach the other end of the edge?

zou3519 replied:
Let's say we have x -> y and we're doing a dfs forward from x. The problem is that in its current state dfs forward is going to try to traverse the x -> y edge and then return immediately (or throw). I could have the dfs forward ignore the x -> y edge, but then it wouldn't really be called a dfs, or I can rewrite contractionProducesCycle in terms of dfs

zou3519 replied:
I rewrote contractionProducesCycle in terms of DFS for better code reuse

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+// Performs a forward DFS.
+// Marks vertices as 'visited' and inserts them into visited.
+// Returns false immediately if upper_bound is reached.
+// Returns true otherwise.
```
The return flag is kind of hard to interpret. I think it would be much clearer if we could have an address of a "forbidden vertex" which shouldn't be reachable (the other endpoint of the inserted edge), and throw from here if we see it.

apaszke replied:
Oh I see that's what `upper_bound` is. Can we just move the throw to this function?

zou3519 replied:
Yeah dfs is doing a few things for the sake of optimization. upper_bound is both an upper bound and the address of a "forbidden node" -- I think I'll separate the two so the logic is clearer.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+// Performs a backward DFS.
+// Marks vertices as 'visited' and inserts them into visited.
+template <typename T>
+void DynamicDAG<T>::dfsBackward(Vertex<T>* producer, size_t lower_bound,
```
I don't think there's any reason why `dfsForward` and `dfsBackward` need to be separate functions. Can't you just make a single DFS and template it over a pointer to member that lets you either access `in_edges` or `out_edges`? The strict check for bound shouldn't hurt the speed too much, and will never trigger in backward anyway.

zou3519 replied:
Hmm that should work

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+      return a->ord < b->ord;
+    }
+  } custom_less;
+  std::sort(delta.begin(), delta.end(), custom_less);
```
Lol can't the last argument be a lambda?

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+    deltaF_ords.begin(), deltaF_ords.end(), std::back_inserter(computed_ords));
+
+  tmp.insert(tmp.end(), deltaB.begin(), deltaB.end());
+  tmp.insert(tmp.end(), deltaF.begin(), deltaF.end());
```
Ugh 4 allocations just to do this merge sort thing (that can be done in-place)!! I guess we can optimize it in the future 😕

apaszke replied:
plus nit: `reserve` on `tmp`

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,388 @@
+  }
+
+  clearVisited(deltaF);
+  clearVisited(deltaB);
```
Noooo this is such a non-trivial place to call `clearVisited` in!! I think the way to go about it would be to make a new struct called `VisitedVector`, which would behave the same as a regular vector, but would set `visited` to `false` on all its elements when it would get destructed.

You should never touch the `visited` attribute manually. It should be set to true on insertion to one of those vectors, and set to false on deletion.

apaszke commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+        // dependency graph (in this case, the Vertex containing the if Node*)
+        // and then record the dependency.
+        auto * owning_node = use.user;
+        while (owning_node != nullptr) {
```
Shouldn't reaching `nullptr` be an error here? We should eventually end up in the current scope. Can you assert on it not happening after the loop? (to make sure we exit via break)

zou3519 replied:
I wasn't sure about how param nodes and return nodes work. Currently the param node and return node of a graph aren't in DynamicDAG (maybe they should be). 

If they aren't in DynamicDAG and if the use.user is the return node, then owning_node can go up to nullptr. I agree that this is weird and I'll try to see if I can figure out what is going on here

zou3519 replied:
True, that sounds reasonable

apaszke commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+
+static void find_differentiable_groups(
+    detail::DynamicDAG<Node*>& dep_graph,
+    size_t distance_threshold=64,
```
You're really generous here 😄

apaszke commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+    detail::DynamicDAG<Node*>& dep_graph,
+    size_t distance_threshold=64,
+    size_t producer_outedge_threshold=16) {
+  // A Vertex contains a Node* or a differential group of Node*.
```
nit: differential -> differentiable

apaszke commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+  int64_t ord = dep_graph.maybe_vertices().size() - 1;
+  while (ord >= 0) {
+    if (!dep_graph.maybe_vertices().at(ord)) {
+      --ord;
```
We decrement this on 3 paths, and want to keep its value on a single path. Maybe let's write this thing as a `for` loop which decrements, and simply do `++ord` in that single place that actually wants to stay at the same index (with a comment)?

apaszke commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+    // Iterate through consumer->in_edges in reverse topological order
+    detail::vertex_list<Node*> in_edges;
+    in_edges.assign(consumer->in_edges.begin(), consumer->in_edges.end());
+    dep_graph.sort(in_edges);
```
This looks like an argument for keeping adjacency lists as sorted vectors (complexity wouldn't be worse than with sets, but we'd have better data locality).

apaszke commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+      // If the edge contraction is successful, consumer->in_edges
+      // may have changed so we break out of this loop.
+      if (dep_graph.contractEdge(producer, consumer)) {
+        changed = true;
```
Think about how many times can we keep re-sorting the `in_edges`!

zou3519 replied:
I've changed it so that we are now storing in_edges in sorted order so we don't keep on sorting it.

apaszke commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -117,11 +123,165 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+
+    auto& rnodes = it->value()->rdata;
+    for (auto it = rnodes.rbegin(); it != rnodes.rend(); ++it) {
+      (*it)->moveBefore(block->return_node());
```
Lol this is a bit surprising at the end, but it makes sense once you realize that we'll move all nodes even if it was unnecessary. A comment might be useful

apaszke reviewed (commented):
Looks nice, but I think the code could be made a bit cleaner and safer.

zou3519 committed:
switch to std::unique_ptr

zou3519 committed:
Address most comments:

- Refactor in_edges/out_edges to helper struct
- Refactor dfsforward & dfsbackward to one dfs(...) fn
- RAII visited_list
- Use vector instead of unordered_set to represent edges
- Separate test cases
- Other minor cleanups

zou3519 committed:
add heuristic to choose between consumer->producer merge or consumer<-producer merge

zou3519 committed:
reorder() now uses 2 vector allocations instead of 5

zou3519 committed:
Remove one vector allocation

zou3519 committed:
Keep adjacency lists in sorted order.

This should not affect complexity too much because the
assumption is that our graphs will not fan-in or fan-out too much.

zou3519 committed:
Switch rdata to data (don't store in reverse order anymore)

apaszke commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -117,11 +145,167 @@ void CreateAutodiffSubgraphs(Block * block, size_t threshold, std::vector<Node*>
+        changed = true;
+        break;
+      }
+    }
```
The `changed` flag is really unnecessary. You only set it in one place, so you could have just as easily done `++ord`

zou3519 commented on test/test_jit.py:
```diff
@@ -8023,6 +8023,163 @@ def clone_inputs(requires_grad):
+        graph = self._perform_ad_subgraph_slicing(fn, 1, 1)
+
+        # We should not have combined the two multiplications into
+        # the same group; they should each be a separate DiffGraph
```
y is used inside the `if bool(k)`, so if we moved `y` to right before `z` we get the following graph:
```
        def fn(x, k):
            if bool(k):
                k += y  # y is out of scope!
            y = x * 1.1
            z = y * k
            return z, k
```

zou3519 commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,534 @@
+
+  static bool has(vertex_list<T>& lst, Vertex<T>* v) {
+    return std::find(lst.begin(), lst.end(), v) != lst.end();
+  }
```
Sounds good!

zou3519 replied:
So actually my current implementation of the sorted_vertex_list is incorrect because `reorder(...)` can shift vertices around such that adjacency lists are no longer in order...

Going to think about this for a while

zou3519 replied:
Resolved, I wrote a wrapper struct

zou3519 commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,534 @@
+  bool visited() { return visited_; }
+
+private:
+  friend DynamicDAG<T>;
```
good point

zou3519 commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,534 @@
+  for (; inputF != deltaF.end(); ++inputF) {
+    *output = (*inputF)->ord;
+    ++output;
+  }
```
ord(y) > deltaB ords, deltaF ords > ord(x). 

The algorithm does std::merge: it knows that we've already sorted deltaB and deltaF individually and now need to recombine it. I could call std::merge but that would involve one extra allocation. Should I refactor this code to a helper function instead?

apaszke replied:
What if we concatenated storages of those vectors and used `std::inplace_merge`?

zou3519 replied:
Ah good idea, didn't know `std::inplace_merge` was a think. Thanks!

zou3519 committed:
Fix sorted assumption

zou3519 committed:
Fix nits

zou3519 committed:
Bound sorting time in algorithm

zou3519 committed:
Make merge operation clearer

zou3519 committed:
Delete some dead code

zou3519 commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,534 @@
+  }
+  ss << "]\n";
+  return ss.str();
+}
```
I deleted the std::string Vertex<Node*>::toString() specialization and kept this one.

zou3519 committed:
Split out helper functions from EdgeData

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,523 @@
+  }
+
+  static void erase(vertex_list<T>& lst, Vertex<T>* v) {
+    lst.erase(std::find(lst.begin(), lst.end(), v));
```
I don't think this code is correct. Those adjacency lists should effectively be sets, because we don't care about multi-edges, but what you have in here is effectively a multi-set.

apaszke replied:
Given that we can't represent them as sorted vectors, I really think we might want to revert the changes back to a regular `unordered_set`. Sorry for all the noise 😕

zou3519 replied:
addEdges(x, y) doesn't add an edge from x to y if it already exists (DynamicDAG is used to build a dependency graph that is a simpler version of the values graph -- I should write this somewhere in a comment). Because of this, these adjacency lists should contain unique elements.

No problem, it's not hard to switch it back to `unordered_set`. I remember that one of the reasons for `vector` over `unordered_set` was that it would have better performance because of data locality and that the collections would be small. Given that, what are the benefits of using `unordered_set` instead?

apaszke replied:
I’m ok with using vectors, but we at least should clearly indicate that this data structure has multi set semantics, and why is that ok in the context of this algorithm.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,534 @@
+ * This is O(|AR|) which is the same thing as O(ord(consumer) - ord(producer))
+ * AR is the "affected region": { v s.t. ord(v) in [ord(producer), ord(consumer)] }
+ * consisting of the only verticies that can possibly be moved around due
+ * to this edge addition.
```
It would be nice if we could say how the reordering happens, because it's a fairly non-trivial part of the algorithm (although very simple once you understand it).

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,523 @@
+    desired_vertex_ordering[i]->ord = sorted_ords[i];
+    vertices_[sorted_ords[i]] = std::move(desired_vertex_ordering[i]);
+  }
+}
```
I don't think this is the same algorithm as the one used to reorder vertices in the paper. I sent you a private message to discuss.

zou3519 replied:
Discussed offline, the algorithm should be the same as the one in the paper

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,523 @@
+
+  // Sort the ords by merging two already sorted lists into a large sorted list.
+  // input (example): deltaB = { v(1), v(4), v(7) } , deltaF = { v(0), v(2), v(5) }.
+  // output: { 0, 1, 2, 5, 7 }.
```
looks like you've missed the 4 in here

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,523 @@
+
+  // Use deltaB to hold the concatenation of deltaB and deltaF.
+  auto middle = deltaB.size();
+  deltaB.insert(deltaB.end(), deltaF.begin(), deltaF.end());
```
Ugh mutating them like that is not perfect since you're taking in references. Just make the arguments values and move them into this function

apaszke replied:
Alternatively, push ords of deltaB and deltaF into sorted_ords and do `inplace_merge` on it!

zou3519 replied:
sounds good!

zou3519 committed:
Address comments:

- Explain how reordering occurs in example graph
- Use ordered_vertex_set for adjacency lists. It is a wrapper around std::vector.
- Add comments that DynamicDAG is a simple graph (has no multi-edges).
  Adjacency lists should not be multi-sets.
- Don't mutate deltaF / deltaB, move them into the reorder() function.
- perform std::inplace_merge on list of ords rather than vertices to
  make the code clearer.

zou3519 committed:
Fix comments about complexity of the algorithms

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,561 @@
+template <typename T> using vertex_list = std::vector<Vertex<T>*>;
+template <typename T> using unique_vertex = std::unique_ptr<Vertex<T>>;
+
+enum DFSDirection {forward, backward};
```
nit: it's better to use `class enum`, especially that this is a header file

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -0,0 +1,561 @@
+// In all of the complexity analysis it is assumed that
+// inserting, erasing, and finding take constant time.
+template <typename T>
+struct ordered_vertex_set {
```
What's ordered about it? I thought we don't give any guarantees.

apaszke replied:
I'd really much rather call it a `multiset`

zou3519 replied:
It's ordered in the sense that there is some ordering to its elements (it matters what order you iterate through the elements).

ordered_vertex_set isn't used as a multiset -- I can add asserts to insert, erase, and contains so that this data structure doesn't get used as a multiset

apaszke replied:
Not being used as a multiset doesn't mean that this data structure doesn't have multiset semantics (this is very important!!).

I wouldn't rely on the ordering all that much since it can be arbitrary unless you call `sort()` before you iterate over it.

zou3519 replied:
Ah I see. Right now, this data structure doesn't completely have multiset semantics: `erase()` is implemented incorrectly for a multiset because it just erases one element and not all elements, so I am wary of naming it something that it is not.

I'd like this data structure to not have multi-set semantics so I think I will add checks on the insert for uniqueness.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -265,8 +280,22 @@ EdgeData<T> DynamicDAG<T>::removeVertex(Vertex<T>* v) {
+ *
+ * 3) Merge the sorted ords: R = { 1, 2, 3, 4, 5, 6 }.
+ *
+ * 4) Reassign the vertices with the sorted ords.
```
The reassignment details are important too! Just add a sentence saying that we always use vertices of deltaB, then x, then y and finally deltaF.

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -426,7 +458,7 @@ bool DynamicDAG<T>::dfsSearch(
     stack.pop_back();
 
-    vertex_list<T>& next_edges = (direction == DFSDirection::forward) ?
+    ordered_vertex_set<T>& next_edges = (direction == DFSDirection::forward) ?
```
you could just use `auto &` in here

apaszke commented on torch/csrc/jit/dynamic_dag.h:
```diff
@@ -447,41 +479,43 @@ bool DynamicDAG<T>::dfsSearch(
-  sort(deltaB);
-  sort(deltaF);
+void DynamicDAG<T>::reorder(visited_list<T> deltaF, visited_list<T> deltaB) {
+  auto deltaB_ = deltaB.vector();
```
can you move those vectors? It will be a copy otherwise.

zou3519 replied:
Do you mean rewrite this as `reorder(visited_list<T>&& deltaF, visited_list<T>&& deltaB)`? (Sorry I am not very good with rvalues).

Right now I am assuming that the compiler will move those values for me (I'm calling `reorder(std::move(deltaF), std::move(deltaB)`) but that is probably not correct...

zou3519 replied:
oh of course! thank you for the explanation. Making deltaB.vector() return a `const std::vector<Vertex<T>*>&` sounds like the way to go

zou3519 committed:
Address comments:

- use 'enum class' for DFSDirection
- rename ordered_vertex_set to vertex_set and enforce set semantics.
  There is no longer multi-set semantics.
- Add one more line about reassigning details

zou3519 committed:
visited_list::vector() now returns const ref

zou3519 committed:
fix unmatched parenthesis
